### PR TITLE
[multistage] [feature] Support Right join and Full join and inEqui mix.

### PR DESCRIPTION
**Pull request by 61yao**
Add full join and right join support. 

This PR assumes that the broadcast partitioned right join table can fit in memory.
This PR also modifies the test framework to allow specifying empty output to skip h2 comparison. (needed since h2 doesn't support full outer join) 

This PR also fixes the bug for left join and inequi join:
https://github.com/apache/pinot/issues/9986

### CODE REVIEW

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java [hunk 1]
```diff
@@ -146,10 +146,18 @@ public DataBlock.Type getType() {
...
   public boolean isEndOfStreamBlock() {
     return isType(MetadataBlock.MetadataBlockType.ERROR) || isType(MetadataBlock.MetadataBlockType.EOS);
   }
 
+  /**
+   * @return true when the block is a real end of stream block instead of error block.
+   */
+  public boolean isEOSBlock() {
+    return isType(MetadataBlock.MetadataBlockType.EOS);
+  }
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java:159]: IMO, simply renamed this to 
`isSuccessfulEndOfStreamBlock()` should be suffice. makes it less confusing and no need to a refactoring.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 2]
```diff
@@ -42,48 +41,70 @@
...
+ * We currently support left join, inner join, right join and full join.
  * The output is in the format of [left_row, right_row]
  */
 // TODO: Move inequi out of hashjoin. (https://github.com/apache/pinot/issues/9728)
 public class HashJoinOperator extends BaseOperator<TransferableBlock> {
+  private class BroadcastRows {
+    public BroadcastRows(List<Object[]> rows, boolean hasMatch) {
+      _rows = rows;
+      _hasMatch = hasMatch;
+    }
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:60]: i am not sure what's the goal here. but higher level question is do we plan to support inequality outer join conditions. 
1. if we do not (e.g. only equality outer join): then there's no need to encapsulate these, simply create another hashmap might be better. 
    - if the join type is LEFT, there' no need to maintain this boxed object and the other HashMap<Key, Boolean> will be empty
    - if the join type is RIGHT/FULL the sole addition is on the HashMap
2. if we want to support inequality join condition, then the hasMap should be attached to every row otherwise the results wont be correct.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 3]
```diff
@@ -42,48 +41,70 @@
...
   private KeySelector<Object[], Object[]> _leftKeySelector;
   private KeySelector<Object[], Object[]> _rightKeySelector;
 
   public HashJoinOperator(Operator<TransferableBlock> leftTableOperator, Operator<TransferableBlock> rightTableOperator,
-      DataSchema outputSchema, JoinNode.JoinKeys joinKeys, List<RexExpression> joinClauses, JoinRelType joinType) {
-    Preconditions.checkState(SUPPORTED_JOIN_TYPES.contains(joinType),
-        "Join type: " + joinType + " is not supported!");
-    _leftKeySelector = joinKeys.getLeftJoinKeySelector();
-    _rightKeySelector = joinKeys.getRightJoinKeySelector();
+      DataSchema leftSchema, JoinNode node) {
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:88]: why is this change necessary?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 4]
```diff
@@ -148,39 +172,73 @@ private TransferableBlock buildJoinedDataBlock(TransferableBlock leftBlock)
...
       _upstreamErrorBlock = leftBlock;
       return _upstreamErrorBlock;
-    } else if (TransferableBlockUtils.isNoOpBlock(leftBlock) || TransferableBlockUtils.isEndOfStream(leftBlock)) {
+    }
+    if (leftBlock.isNoOpBlock() || (leftBlock.isEOSBlock() && (_joinType != JoinRelType.FULL
+        && _joinType != JoinRelType.RIGHT))) {
       return leftBlock;
     }
+    if (leftBlock.isEOSBlock()) {
+      // Return remaining non-matched rows for full join in broadcast table
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:181]: return remaining non-matched rows for full or right OUTER joins

**Review by walterddr — COMMENTED** [review]
could you please rebase and trigger CI?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java (hunk 2 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:60]: alternatively we can also maintain 2 hashmaps
```
private final HashMap<Key, List<Object[]>> _unmatchedHashTable;
private final HashMap<Key, List<Object[]>> _matchedHashTable;
```
and for the hash join, simply loop through both 
1. for LEFT and INNER join, remain the same logic (the _matchHashTable will always be empty and thus no extra cost)
2. for RIGHT and FULL join, move a row from _unmatch to _match onc the equality and inequality conditions are applied successfully. 
    - at the very end when a EOS block is reached on the left. simply output the right _umatched rows

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java (hunk 1 above)
Comment (61yao) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java]: Done

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java (hunk 2 above)
Comment (61yao) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java]: This doesn't work since left join has a bug when there is inequi condition. we have to record _unmatchedLeftRows for left join and full join.

For right join and full join, we have to remember matched rows to output unmatched ones.
I used  HashMap<Key, HashSet<Integer>> _matchedRightRows because we have to use this to look up unmatched rows.

PTAL at the new implementation.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java (hunk 3 above)
Comment (61yao) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:108]: not necessary. but I made the change to make the ctor simpler.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java (hunk 4 above)
Comment (61yao) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java]: Updated the comment

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java (hunk 3 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:108]: +1 got it

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 5]
```diff
@@ -43,50 +49,94 @@
...
+    return _joinType == JoinRelType.RIGHT || _joinType == JoinRelType.FULL;
+  }
+
+  private boolean needUnmatchedLeftRows() {
+    return _joinType == JoinRelType.LEFT || _joinType == JoinRelType.FULL;
+  }
+
+  private boolean needUnmatchedRows() {
+    return _joinType != JoinRelType.INNER;
+  }
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:105]: nit: private util methods are usually put after the public ones in pinot. so please move these to the end of file

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 6]
```diff
@@ -152,39 +202,82 @@ private TransferableBlock buildJoinedDataBlock(TransferableBlock leftBlock)
     if (leftBlock.isErrorBlock()) {
       _upstreamErrorBlock = leftBlock;
       return _upstreamErrorBlock;
-    } else if (TransferableBlockUtils.isNoOpBlock(leftBlock) || TransferableBlockUtils.isEndOfStream(leftBlock)) {
+    }
+    if (leftBlock.isNoOpBlock() || (leftBlock.isSuccessfulEndOfStreamBlock() && !needUnmatchedRows())) {
       return leftBlock;
     }
+    if (leftBlock.isSuccessfulEndOfStreamBlock()) {
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:209]: nit: suggest putting the logic of the end of stream handling into a separate function for better readability.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 7]
```diff
@@ -152,39 +202,82 @@ private TransferableBlock buildJoinedDataBlock(TransferableBlock leftBlock)
...
-          // TODO: Optimize this to avoid unnecessary object copy.
-          Object[] resultRow = joinRow(leftRow, rightRow);
-          if (_joinClauseEvaluators.isEmpty() || _joinClauseEvaluators.stream().allMatch(evaluator ->
-              (Boolean) FunctionInvokeUtils.convert(evaluator.apply(resultRow), DataSchema.ColumnDataType.BOOLEAN))) {
-            rows.add(resultRow);
+      List<Object[]> matchedRightRows = _broadcastRightTable.getOrDefault(key, null);
+      if (matchedRightRows == null) {
+        if (needUnmatchedLeftRows()) {
+          rows.add(joinRow(leftRow, null));
+        }
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:244]: why do we do this here. and then again in line 224-228?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 8]
```diff
@@ -152,39 +202,82 @@ private TransferableBlock buildJoinedDataBlock(TransferableBlock leftBlock)
...
+      if (needUnmatchedRightRows()) {
+        for (Map.Entry<Key, List<Object[]>> entry : _broadcastRightTable.entrySet()) {
+          Set<Integer> matchedIdx = _matchedRightRows.getOrDefault(entry.getKey(), new HashSet<>());
+          int rowSize = entry.getValue().size();
+          Set<Integer> allIdx = IntStream.range(0, rowSize).boxed().collect(Collectors.toSet());
+          Set<Integer> unmatchedIdx = Sets.difference(allIdx, matchedIdx);
+          for (int unmatchedRow : unmatchedIdx) {
+            Object[] resultRow = joinRow(null, entry.getValue().get(unmatchedRow));
+            returnRows.add(resultRow);
+          }
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:221]: can we avoid using stream? this is on the critical code path (for every hashkey) so it is better to do a single loop

```suggestion
          Set<Integer> matchedIdx = _matchedRightRows.getOrDefault(entry.getKey(), new HashSet<>());
          List<Object[]> rightRows = entry.getValues();
          for (int i = 0; i <  rightRows.size(); i++) {
            if (!matchedIdx.contains(i)) {
              Object[] resultRow = joinRow(null, rightRows.get(i));
              returnRows.add(resultRow);
            }
          }
```

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java (hunk 7 above)
Comment (61yao) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:244]: Actually, it is an issue caused by my previous implementation. I updated this

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java (hunk 8 above)
Comment (61yao) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java]: Done

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java (hunk 7 above)
Comment (61yao) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:214]: Updated. PTAL

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java (hunk 6 above)
Comment (61yao) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java]: I added a TODO to follow up.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java (hunk 5 above)
Comment (61yao) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java]: done.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 9]
```diff
@@ -18,12 +20,13 @@
  */
 package org.apache.pinot.query.runtime.operator;
 
-import com.google.common.base.Preconditions;
+import com.clearspring.analytics.util.Preconditions;
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:23]: do not change this

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 10]
```diff
@@ -152,44 +179,87 @@ private TransferableBlock buildJoinedDataBlock(TransferableBlock leftBlock)
     if (leftBlock.isErrorBlock()) {
       _upstreamErrorBlock = leftBlock;
       return _upstreamErrorBlock;
-    } else if (TransferableBlockUtils.isNoOpBlock(leftBlock) || TransferableBlockUtils.isEndOfStream(leftBlock)) {
+    }
+    if (leftBlock.isNoOpBlock() || (leftBlock.isSuccessfulEndOfStreamBlock() && !needUnmatchedRightRows())) {
       return leftBlock;
     }
+    // TODO: Moved to a different function.
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:186]: +1 thanks for recording this

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 11]
```diff
@@ -152,44 +179,87 @@ private TransferableBlock buildJoinedDataBlock(TransferableBlock leftBlock)
...
-          if (_joinClauseEvaluators.isEmpty() || _joinClauseEvaluators.stream().allMatch(evaluator ->
-              (Boolean) FunctionInvokeUtils.convert(evaluator.apply(resultRow), DataSchema.ColumnDataType.BOOLEAN))) {
-            rows.add(resultRow);
+      List<Object[]> matchedRightRows = _broadcastRightTable.getOrDefault(key, null);
+      if (matchedRightRows == null) {
+        if (needUnmatchedLeftRows()) {
+          rows.add(joinRow(leftRow, null));
+        }
+        continue;
+      }
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:216]: suggest clean up for the logic. no need to add join rows separately for equality or inequality in 2 places
```suggestion
      boolean hasMatchForLeftRow = false;
      if (matchedRightRows != null) {
        // ...
      } 
      if (!hasMatchForLeftRow && needUnmatchedLeftRows()) {
        rows.add(joinRow(leftRow, null));
      }
```

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 12]
```diff
@@ -152,44 +179,87 @@ private TransferableBlock buildJoinedDataBlock(TransferableBlock leftBlock)
...
-    for (Object obj : leftRow) {
-      resultRow[idx++] = obj;
+    if (leftRow != null) {
+      for (Object obj : leftRow) {
+        resultRow[idx++] = obj;
+      }
     }
+    // This is needed since left row can be null and we need to advance the idx to the beginning of right row.
+    idx = _leftRowSize;
     if (rightRow != null) {
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:250]: suggest making this into 3 utils. but can be done in follow up

**Review by walterddr — APPROVED** [review]
lgtm. other than the import change the rest of the comments can be address separately. 

thank you for addressing the performance issues

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java (hunk 9 above)
Comment (61yao) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java]: Good catch. I think it is an accidental change... I am not sure how it happens. updated.

**Comment by walterddr** [comment]
merged. please follow up with the perf improvement comments in separate PR. thank you @61yao
